test/e2e/network/endpointslice.go

```diff
@@ -299,6 +300,113 @@ var _ = SIGDescribe("EndpointSlice", func() {
+		// endpoints. Not waiting for all Pods to be part of EndpointSlice to
+		// ensure EndpointSlice controller is even busier in following step.
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		err := wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+			esList, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+			framework.ExpectNoError(err, "Error listing EndpointSlices for %s/%s Service", f.Namespace.Name, svc.Name)
+			if len(esList.Items) != 1 {
+				framework.Logf("Expected 1 EndpointSlice for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items))
+				return false, nil
+			}
+
+			if len(esList.Items[0].Endpoints) < 50 {
+				framework.Logf("Expected at least 50 EndpointSlice endpoints for %s/%s Service, got %d", f.Namespace.Name, svc.Name, len(esList.Items[0].Endpoints))
+				return false, nil
+			}
+
+			return true, nil
+		})
+		framework.ExpectNoError(err, "Timed out waiting for EndpointSlice to have at least 50 endpoints for %s/%s Service", f.Namespace.Name, svc.Name)
+
+		// In this step we run 4 goroutines simultaneously. The first 3 are all
+		// trying to keep the EndpointSlice controller as busy as possible.
+		// There are 2 goroutines creating 51 new Pods each at the same time
+		// another goroutine is deleting the 100 Pods created initially. The
+		// final goroutine deletes the Service in the midst of the chaos to
+		// ensure that all of this action in the EndpointSlice controller will
+		// not interfere with garbage collection.
+		var wg sync.WaitGroup
+		wg.Add(4)
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 100, 151, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			createTestPods(podClient, 200, 251, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			deleteTestPods(podClient, 0, 100, sharedLabels)
+		}()
+
+		go func() {
+			defer wg.Done()
+			// Wait for at least 2 EndpointSlices to exist for this Service.
+			err := wait.Poll(5*time.Second, 3*time.Minute, func() (bool, error) {
+				esList, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
+				framework.ExpectNoError(err, "Error listing EndpointSlices for %s/%s Service", f.Namespace.Name, svc.Name)
+				if len(esList.Items) < 2 {
+					framework.Logf("Waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+					return false, nil
+				}
+
+				return true, nil
+			})
+			framework.ExpectNoError(err, "Timed out waiting for second EndpointSlice to be created for %s/%s Service", f.Namespace.Name, svc.Name)
+
+			err = cs.CoreV1().Services(f.Namespace.Name).Delete(context.TODO(), svc.Name, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "Error deleting %s/%s Service", f.Namespace.Name, svc.Name)
+		}()
+
+		wg.Wait()
+
+		// Wait for EndpointSlices to be garbage collected.
```

Comment:
When using `metav1.DeletePropagationForeground`, a simpler test is to wait for the service to be removed after marking it for deletion.

Reply:
Thanks! I've added an e2e test that uses this approach.